File: Java/Eclipse/Experiments/src/Vector.java
import java.util.Random;

public class Vector {

	public double x, y, magnitude;

	public Vector(double x, double y) {
		this.x = x;
		this.y = y;
		calcMagnitude();
	}

	public Vector() {
		x = 1;
		y = 0;
		double randomAngle = 2 * Math.PI * (new Random()).nextDouble();
		rotate(randomAngle);
	}

	private void rotate(double angle) {
		double sin = Math.sin(angle);
		double cos = Math.cos(angle);
		double newX = x * cos - y * sin;
		double newY = x * sin + y * cos;
		this.x = newX;
		this.y = newY;
		calcMagnitude();
	}

	private void calcMagnitude() {
		this.magnitude = Math.sqrt(x * x + y * y);
	}

	public Vector add(Vector vector) {
		return new Vector(x + vector.x, y + vector.y);
	}

	public Vector subtract(Vector vector) {
		return new Vector(x - vector.x, y - vector.y);
	}

	public Vector multiply(double factor) {
		return new Vector(x * factor, y * factor);
	}

	public void setMagnitude(double newMagnitude) {
		if(magnitude ==0){
			System.out.println("ERROR");
			return;
		}
		double temp = newMagnitude / magnitude;
		x *= temp;
		y *= temp;
		calcMagnitude();
	}

}

File: JavaScript/MyWork/vector.js
function Vector(x, y){
	this.x = x;
	this.y = y;
	this.magnitude = Math.sqrt(x*x + y*y);

	this.updateMagnitude = function(){
		this.magnitude = Math.sqrt(x*x + y*y);
	}
	this.add = function(vector){
		this.x += vector.x;
		this.y += vector.y;
		this.updateMagnitude();
	}
	this.sub = function(vector){
		this.x -= vector.x;
		this.y -= vector.y;
		this.updateMagnitude();
	}
	this.setMagnitude = function(newMagnitude){
		this.x *= newMagnitude/this.magnitude;
		this.y *= newMagnitude/this.magnitude;
		this.magnitude = newMagnitude;
	}
	this.multiply = function(amount){
		this.x *= amount;
		this.y *= amount;
		this.updateMagnitude();
	}
	this.rotate = function(angle){
		var s = Math.sin(angle);
		var c = Math.cos(angle);
		var newX = (this.x)*c - (this.y)*s;
		var newY = (this.x)*s + (this.y)*c;
		this.x = newX;
		this.y = newY;
	}
	
}
function copy(vector){
	var duplicate = new Vector(vector.x, vector.y);
	return duplicate;
}
function randomVector(){
	var vector = new Vector(1, 0);
	var randomAngle = Math.random()*2*Math.PI;
	vector.rotate(randomAngle);
	return vector;
}